### PR TITLE
Fix: Update grad norm calculation for CPU offload

### LINKED ISSUE #7292: Question about `safe_set_local_grad` and `safe_get_local_grad`
deepspeed==0.16.7
I'm using zero stage-3

I write a custom Trainer to do some custom grad clipping after `backward()` is called and before `step()`. I use `safe_get_local_grad` and `safe_set_local_grad` to do the custom grad clipping. After clipping, I call `safe_get_local_grad` again, and the result is as expected. But then i add a callback on the end of `step()` and print the `ds_engine.optimizer._global_grad_norm`, it equals to the original grad(before the clipping). Does it mean I fail to set new value to grad?

### PR DESCRIPTION
**Pull request by therealnaveenkamal**
## Description
This PR fixes an issue where gradient clipping modifications are not reflected in the global gradient norm calculation when CPU offloading is enabled. The issue occurs because the `averaged_gradients` are not being updated with the clipped gradients when CPU offloading is active.

## Problem
When using CPU offloading with gradient clipping:
1. The gradients are successfully clipped using `safe_set_local_grad`
2. However, the `_global_grad_norm` calculation still uses the original unclipped gradients.
3. This leads to incorrect gradient norm reporting and potential issues with gradient clipping effectiveness

## Solution
The fix ensures that the `averaged_gradients` are properly updated with the clipped gradients when CPU offloading is enabled, similar to how it works when CPU offloading is disabled.

## Testing
The fix has been tested with:
- CPU offloading enabled and disabled
- Different gradient clipping values
- A simple model with linear layers
- Both FP16 and BF16

## Related Issues
Fixes #7292

### CODE REVIEW

**Comment by therealnaveenkamal** [comment]
@sfc-gh-truwase I agree. I've reverted my changes and added a norm update in the local API.

```
if self.offload_optimizer:
            self.norm_for_param_grads[self.get_param_id(param)] = self._constant_buffered_norm2(value)
```

**Comment by sfc-gh-truwase** [comment]
> @sfc-gh-truwase I agree. I've reverted my changes and added a norm update in the local API. Please let me know your thoughts
> 
> ```
> if self.offload_optimizer:
>             self.norm_for_param_grads[self.get_param_id(param)] = self._constant_buffered_norm2(value)
> ```

@therealnaveenkamal, looks good to me. Did you get a chance to test? Also, is it possible to convert your test case into a unit test?

**Comment by therealnaveenkamal** [comment]
@sfc-gh-truwase I've added a unit-test file [test_zero_grad_clip.py](https://github.com/deepspeedai/DeepSpeed/pull/7302/files#diff-a21df099b3831f13b25b6ccb5eca9ee2fe4c7de5d2672356286aeb026c7866aa). Runs fours tests with BF16 and FP16.

File: tests/unit/runtime/zero/test_zero_grad_clip.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+                
+            grad = safe_get_local_grad(param)
+            if grad is not None:
+                pre_clip_norm = grad.norm().item()
+                # Apply clipping
+                clamped_grad = torch.clamp(grad, -clip_value, clip_value)
+                post_clip_norm = clamped_grad.norm().item()
+                
+                if pre_clip_norm > clip_value:
+                    print(f"DEBUG: Param {param.ds_id} - Pre-clip norm: {pre_clip_norm:.6f}, Post-clip norm: {post_clip_norm:.6f}")
```
Comment (sfc-gh-truwase) [inline tests/unit/runtime/zero/test_zero_grad_clip.py:96]: Please remove the `print()` so the unit test is not noisy

Comment (therealnaveenkamal) [reply tests/unit/runtime/zero/test_zero_grad_clip.py]: @sfc-gh-truwase Thanks for letting me know. I've updated the file.

Comment (therealnaveenkamal) [reply tests/unit/runtime/zero/test_zero_grad_clip.py]: @sfc-gh-truwase Looks like the test is failing...module not found error: mpi4py. Can you please help me here?

Comment (sfc-gh-truwase) [reply tests/unit/runtime/zero/test_zero_grad_clip.py]: I will take a look on the next run. Usually, `mpi4py` should have been handled by the requirements.txt. 

Also, did you run the formatting checks? 
https://github.com/deepspeedai/DeepSpeed/blob/master/CONTRIBUTING.md#prerequisites

Comment (therealnaveenkamal) [reply tests/unit/runtime/zero/test_zero_grad_clip.py]: @sfc-gh-truwase I updated my script, works without mpi4py. Also, did the formatting tests. I did sign-off, but still the DCO shows it failed.

and sorry for the trouble caused, I'm a first time contributor!

Comment (therealnaveenkamal) [reply tests/unit/runtime/zero/test_zero_grad_clip.py]: @sfc-gh-truwase looks like there were issues with torch.distributed setup. When I test locally, the tests passed. I've handled exceptions and pushed my updates. hopefully, the workflow passes this time.


```
configfile: pytest.ini
plugins: forked-1.6.0
collected 4 items                                                                                                             

tests/unit/runtime/zero/test_zero_grad_clip.py::TestZeroGradClip::test_grad_clip_and_norm_update[fp16-0.5-cpu] PASSED   [ 25%]
tests/unit/runtime/zero/test_zero_grad_clip.py::TestZeroGradClip::test_grad_clip_and_norm_update[bf16-0.05-cpu] PASSED  [ 50%]
tests/unit/runtime/zero/test_zero_grad_clip.py::TestZeroGradClip::test_grad_clip_and_norm_update[fp16-0.5-none] PASSED  [ 75%]
tests/unit/runtime/zero/test_zero_grad_clip.py::TestZeroGradClip::test_grad_clip_and_norm_update[bf16-0.05-none] PASSED [100%]
```

Comment (sfc-gh-truwase) [reply tests/unit/runtime/zero/test_zero_grad_clip.py]: Sorry about the flaky CI. I will also keep an eye on it. Thanks!

Comment (therealnaveenkamal) [reply tests/unit/runtime/zero/test_zero_grad_clip.py]: @sfc-gh-truwase Thanks for the support. Would love to contribute more.

**Review by sfc-gh-truwase — APPROVED** [review]
